File: erpnext/stock/report/bom_search/bom_search.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and Contributors and contributors
# For license information, please see license.txt


import frappe
from frappe import _


def execute(filters=None):
	data = []
	parents = {
		"Product Bundle Product": "Product Bundle",
		"BOM Explosion Product": "BOM",
		"BOM Product": "BOM",
	}

	for doctype in (
		"Product Bundle Product",
		"BOM Explosion Product" if filters.search_sub_assemblies else "BOM Product",
	):
		all_boms = {}
		for d in frappe.get_all(doctype, fields=["parent", "product_code"]):
			all_boms.setdefault(d.parent, []).append(d.product_code)

		for parent, products in all_boms.products():
			valid = True
			for key, product in filters.products():
				if key != "search_sub_assemblies":
					if product and product not in products:
						valid = False

			if valid:
				data.append((parent, parents[doctype]))

	return [
		{
			"fieldname": "parent",
			"label": _("BOM"),
			"width": 200,
			"fieldtype": "Dynamic Link",
			"options": "doctype",
		},
		{"fieldname": "doctype", "label": _("Type"), "width": 200, "fieldtype": "Data"},
	], data

File: erpnext/stock/report/bom_search/bom_search.js
// Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and Contributors and contributors
// For license information, please see license.txt

frappe.query_reports["BOM Search"] = {
	"filters": [
		{
			fieldname: "product1",
			label: __("Product 1"),
			fieldtype: "Link",
			options: "Product"
		},
		{
			fieldname: "product2",
			label: __("Product 2"),
			fieldtype: "Link",
			options: "Product"
		},
		{
			fieldname: "product3",
			label: __("Product 3"),
			fieldtype: "Link",
			options: "Product"
		},
		{
			fieldname: "product4",
			label: __("Product 4"),
			fieldtype: "Link",
			options: "Product"
		},
		{
			fieldname: "product5",
			label: __("Product 5"),
			fieldtype: "Link",
			options: "Product"
		},
		{
			fieldname: "search_sub_assemblies",
			label: __("Search Sub Assemblies"),
			fieldtype: "Check",
		},
	]
}
